hpx/lcos/channel.hpp

```diff
@@ -0,0 +1,396 @@
+//  Copyright (c) 2016 Hartmut Kaiser
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#if !defined(HPX_LCOS_CHANNEL_JUL_23_2016_0722PM)
+#define HPX_LCOS_CHANNEL_JUL_23_2016_0722PM
+
+#include <hpx/config.hpp>
+#include <hpx/lcos/server/channel.hpp>
+#include <hpx/lcos/future.hpp>
+#include <hpx/runtime/components/client_base.hpp>
+#include <hpx/runtime/components/new.hpp>
+#include <hpx/runtime/naming_fwd.hpp>
+
+#include <type_traits>
+#include <utility>
+
+namespace hpx { namespace lcos
+{
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename T = void> class channel;
+    template <typename T = void> class receive_channel;
+    template <typename T = void> class send_channel;
+
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename T, typename Channel>
+    class channel_iterator
+      : public hpx::util::iterator_facade<
+            channel_iterator<T, Channel>, T const, std::input_iterator_tag>
+    {
+        typedef hpx::util::iterator_facade<
+                channel_iterator<T, Channel>, T const, std::input_iterator_tag
+            > base_type;
+
+    public:
+        channel_iterator()
+          : channel_(nullptr), data_(T(), false)
+        {}
+
+        explicit channel_iterator(Channel const& c)
+          : channel_(&c), data_(get_checked())
+        {}
+
+    private:
+        std::pair<T, bool> get_checked() const
+        {
+            hpx::future<T> f = channel_->get();
+            f.wait();
+            if (!f.has_exception())
+            {
+                return std::make_pair(f.get(), true);
+            }
+            return std::make_pair(T(), false);
```

Comment:
Yes, this is used to flag the end of a range, so it ignores the exceptions. This is yet another argument in favor of making the iterators lazy.